include/graphqlservice/GraphQLTree.h

```diff
@@ -21,19 +21,129 @@
+	// and never copied or assigned...
+	ast_node(const ast_node&) = delete;
+	ast_node(ast_node&&) = delete;
+	ast_node& operator=(const ast_node&) = delete;
+	ast_node& operator=(ast_node&&) = delete;
+
+	// Must be destructible
+	~ast_node() = default;
+
+	[[nodiscard]] GRAPHQLPEG_EXPORT bool is_root() const noexcept;
+	[[nodiscard]] GRAPHQLPEG_EXPORT position begin() const noexcept;
+	[[nodiscard]] GRAPHQLPEG_EXPORT std::string_view string_view() const noexcept;
+	[[nodiscard]] GRAPHQLPEG_EXPORT std::string string() const noexcept;
+	[[nodiscard]] GRAPHQLPEG_EXPORT bool has_content() const noexcept;
+
+	GRAPHQLPEG_EXPORT void unescaped_view(std::string_view unescaped) noexcept;
+	[[nodiscard]] GRAPHQLPEG_EXPORT std::string_view unescaped_view() const;
+
+	template <typename U>
+	[[nodiscard]] bool is_type() const noexcept
+	{
+		const auto u = type_name<U>();
+		const auto u_hash = type_hash<U>();
+
+		return _type_hash == u_hash
+			&& ((_type.data() == u.data() && _type.size() == u.size()) || _type == u);
+	}
+
+	template <typename... States>
+	void remove_content(States&&...) noexcept
+	{
+		_content = {};
+		_unescaped.reset();
+	}
+
+	// All non-root nodes receive a call to start() when
+	// a match is attempted for Rule in a parsing run...
+	template <typename Rule, typename ParseInput, typename... States>
+	void start(const ParseInput& in, States&&...)
+	{
+		_begin = in.iterator();
+	}
+
+	// ...and later a call to success() when the match succeeded...
+	template <typename Rule, typename ParseInput, typename... States>
+	void success(const ParseInput& in, States&&...)
+	{
+		const char* end = in.iterator().data;
+
+		_type = type_name<Rule>();
+		_type_hash = type_hash<Rule>();
+		_source = in.source();
+		_content = { _begin.data, static_cast<size_t>(end - _begin.data) };
+		_unescaped.reset();
+	}
+
+	// ...or to failure() when a (local) failure was encountered.
+	template <typename Rule, typename ParseInput, typename... States>
+	void failure(const ParseInput&, States&&...)
+	{
+	}
+
+	// if parsing of the rule failed with an exception, this method is called
+	template <typename Rule, typename ParseInput, typename... States>
+	void unwind(const ParseInput&, States&&...) noexcept
+	{
+	}
+
+	// After a call to success(), and the (optional) call to the selector's
+	// transform() did not discard a node, it is passed to its parent node
+	// with a call to the parent node's emplace_back() member function.
+	template <typename... States>
+	void emplace_back(std::unique_ptr<ast_node> child, States&&...)
+	{
+		children.emplace_back(std::move(child));
+	}
+
+	using children_t = std::vector<std::unique_ptr<ast_node>>;
+
+	children_t children;
+
+private:
+	template <typename U>
+	[[nodiscard]] static std::string_view type_name() noexcept
+	{
+		static const std::string_view name { tao::demangle<U>() };
+
+		return name;
+	}
+
+	template <typename U>
+	[[nodiscard]] static size_t type_hash() noexcept
```

Comment:
there is no need to use this part anymore, the `((_type.data() == u.data() && _type.size() == u.size())` that the PEGTL dev added upstream is enough.

Saves 8 bytes per node :-)

Reply:
I just tweaked that and added a [comment](https://github.com/wravery/cppgraphqlgen/blob/d6d373b89de10a76aefdadb073b006877ba7fb26/include/graphqlservice/GraphQLTree.h#L56-L59). The short version is that pointer comparisons aren't reliable when using shared libraries instead of static libraries. Each module gets its own copy of any static data, and in the case of Windows DLLs, they also get their own copies of any template specializations which are not explicitly exported (shared libraries on Linux don't seem to require an explicit export, so they might share the template specializations).

Reply:
indeed, but people using static linking will benefit.

would also help if we generate a single dynamic lib, instead of various. Then peg/service would be in the same, all using the same pointer.

the `std::hash` is not that helpful since we'd be computing the to-be-compared type's hash, thus walking the whole string at least once. In that case, `strcmp()` would stop at the first non-matching character...

Reply:
That's why I put it in a static local variable in the `type_hash` template method. It should lazy init the hash value once per template instantiation. The only per-node cost to enable this for either static or shared libraries should be the 4 - 8 byte member which is set in the `success` callback.